utils/pattern/src/replacement.rs

```diff
@@ -0,0 +1,203 @@
+///     type Item = Element;
+///
+///     fn next(&mut self) -> Option<Self::Item> {
+///         self.idx += 1;
+///         if self.idx <= self.max {
+///             Some(Element::Digit(self.idx))
+///         } else {
+///             None
+///         }
+///     }
+/// }
+///
+///
+/// impl ReplacementProvider<usize, Element> for MyReplacementProvider {
+///     type Iter = MyIterator;
+///
+///     fn take_replacement(&mut self, key: &usize) -> Option<Self::Iter> {
+///         Some(MyIterator { idx: 0, max: *key })
+///     }
+/// }
+///
+/// let mut parser = Parser::new("{4}, {2}");
+/// let mut interpolator = Interpolator::new(parser, MyReplacementProvider);
+///
+/// assert_eq!(Ok(Some(Element::Digit(1))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(2))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(3))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(4))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Literal(", ".to_string()))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(1))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(2))), interpolator.try_next());
+/// assert_eq!(Ok(None), interpolator.try_next());
+/// ```
+pub trait ReplacementProvider<K, E> {
+    type Iter: Iterator<Item = E>;
+
+    /// Retrieves a replacement iterator to be used by the [`Interpolator`] in
+    /// place of a placeholder.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::ReplacementProvider;
+    /// use std::collections::HashMap;
+    ///
+    /// #[derive(Debug, PartialEq)]
+    /// enum Element {
+    ///     TokenFive,
+    /// }
+    ///
+    /// impl ReplacementProvider<usize, Element> for HashMap<usize, Vec<Element>> {
+    ///     type Iter = <Vec<Element> as IntoIterator>::IntoIter;
+    ///
+    ///     fn take_replacement(&mut self, key: &usize) -> Option<Self::Iter> {
+    ///         let r = self.remove(key)?;
+    ///         Some(r.into_iter())
+    ///     }
+    /// }
+    ///
+    /// let mut replacements = HashMap::new();
+    /// replacements.insert(5, vec![
+    ///     Element::TokenFive
+    /// ]);
+    ///
+    /// assert_eq!(
+    ///     replacements.take_replacement(&5).map(|r| r.collect()),
+    ///     Some(vec![Element::TokenFive])
+    /// );
+    ///
+    /// assert_eq!(
+    ///     replacements.take_replacement(&1).map(|r| r.collect::<Vec<_>>()),
+    ///     None
+    /// );
+    /// ```
+    ///
+    /// [`Interpolator`]: crate::interpolator::Interpolator
+    fn take_replacement(&mut self, key: &K) -> Option<Self::Iter>;
+}
+
+impl<E> ReplacementProvider<usize, E> for Vec<Vec<E>> {
+    type Iter = <Vec<E> as IntoIterator>::IntoIter;
+
+    fn take_replacement(&mut self, input: &usize) -> Option<Self::Iter> {
+        let r = self.get_mut(*input)?;
+        Some(std::mem::take(r).into_iter())
+    }
+}
+
+impl<E> ReplacementProvider<String, E> for HashMap<String, Vec<E>> {
```

Comment:
Suggestion: Seems like a good use case for `TinyStrAuto` instead of `String`.

Reply:
I mostly used it as an example. We don't have a real use case for it yet, and once we do, we can fine tune what types we implement the trait for. I prefer to use `String` for an example, because it generalizes better and reduces dependency.

If you have a particular use case where we'd like to use the pattern combinator with named placeholders, I'd love to see it. From my experience it seems that CLDR/Unicode so far is not doing that - one of my hopes is that by providing such a design we can now reason about switching numbered placeholders to named ones.

I'm a fan of named placeholders because they are more meaningful and handle error cases much better (`{date}` displayed to the user in an error scenario is more meaningful than `{0}`.), but I more designed it this way as future-proof, without a tangible use case right now.
Hence this can be seen more as an example of how it might be used in the future, than an impl that will be used by us right now.